### PR TITLE
Add parens to head of `ExpressionStatement` instead of whole statement

### PR DESCRIPTION
**Pull request by fisker**
## Description

Ref https://github.com/prettier/prettier/issues/13999#issuecomment-1354050144

//cc @Josh-Cena 

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Review by thorn0 — APPROVED** [review]

**Comment by fisker** [comment]
@thorn0 

I saw you are adding `callAncestor`, are worrying about the performance issue? 
I don't think this PR makes much differences.

You may want try use `getAncestors()` added in `next` branch, you can add an option to yield key info to it.

**Comment by thorn0** [comment]
I'm [experimenting](https://github.com/thorn0/prettier/commits/callAncestor-7976) with solving https://github.com/prettier/prettier/issues/7976, and it works, but I'm not sure about the performance of this solution.
The idea of `callAncestor` is to get a path object for the ancestor, so that we could call `needsParens` for it (or any other function that needs a path). `getAncestors()` won't give me a path.

**Comment by thorn0** [comment]
> but I agree this can be improved without searching useless ExpressionStatement

I don't see what you mean here BTW. I just needed to pass a path to `startsWithNoLookaheadToken` instead of a node.

**Comment by fisker** [comment]
I mean in this PR. The following case

```js
a(
	b(
		c(
			{}.toString()
		)
	)
)
```

The `ObjectExpression` is obviously not the head of `ExpressionStatement`, but we can't tell until I found `ExpressionStatement`.

**Comment by fisker** [comment]
> I don't think this PR makes much differences, 

Previously, we can't tell that variable `a` is not a node need parens, until it found the most left node. So I guess there are not much difference.

**Comment by fisker** [comment]
When I learn write babel plugin, I saw babel's [`NodePath`](https://github.com/babel/babel/blob/6e9f7b77151654f59feee7b9bc6ecd420869edae/packages/babel-traverse/src/path/index.ts#L35) is not a state machine. But I haven't read how it works. Maybe we can borrow some ideas from there.

**Comment by thorn0** [comment]
They seem to create a new path object for each node. It's as an optimization that Prettier doesn't do that.

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -128,6 +128,26 @@ function needsParens(path, options) {
     return false;
   }
 
+  if (
+    node.type === "ObjectExpression" ||
+    node.type === "FunctionExpression" ||
+    node.type === "ClassExpression" ||
+    node.type === "DoExpression"
```
Comment (sosukesuzuki) [inline src/language-js/needs-parens.js:135]: `RecordExpression` should be included?

Comment (fisker) [reply src/language-js/needs-parens.js:135]: No, `#{}.toString()` is valid.

**Pull request merged**